--- ld/pe-dll.c	Mon Aug 06 08:32:08 2007
+++ ld/pe-dll.c	Tue Oct 16 18:47:02 2007
@@ -598,10 +598,9 @@
 	}
     }
 
-  /* If we are not building a DLL, when there are no exports
+  /* If there are no exports
      we do not build an export table at all.  */
-  if (!pe_dll_export_everything && pe_def_file->num_exports == 0
-      && info->executable)
+  if (!pe_dll_export_everything && pe_def_file->num_exports == 0)
     return;
 
   /* Now, maybe export everything else the default way.  */
