--- ld/pe-dll.c	Thu Jul 19 13:12:24 2007
+++ ld/pe-dll.c	Fri Aug  3 09:53:58 2007
@@ -524,10 +524,9 @@
 	}
     }
 
-  /* If we are not building a DLL, when there are no exports
+  /* If there are no exports
      we do not build an export table at all.  */
-  if (!pe_dll_export_everything && pe_def_file->num_exports == 0
-      && info->executable)
+  if (!pe_dll_export_everything && pe_def_file->num_exports == 0)
     return;
 
   /* Now, maybe export everything else the default way.  */